runtime/vam/expr/gencomparefuncs.go

```diff
@@ -0,0 +1,316 @@
+			out.Set(k)
+		}
+	}
+`
+
+const flat_const = `
+	for k := uint32(0); k < n; k++ {
+		if l.Values[k] %s rconst {
+			out.Set(k)
+		}
+	}
+`
+
+const idx_flat = `
+	for k := uint32(0); k < n; k++ {
+		if l.Values[lx[k]] %s r.Values[k] {
+			out.Set(k)
+		}
+	}
+`
+
+const idx_idx = `
+	for k := uint32(0); k < n; k++ {
+		if l.Values[lx[k]] %s r.Values[rx[k]] {
+			out.Set(k)
+		}
+	}
+`
+
+const idx_const = `
+	for k := uint32(0); k < n; k++ {
+		if l.Values[lx[k]] %s rconst {
+			out.Set(k)
+		}
+	}
+`
+
+const const_flat = `
+	for k := uint32(0); k < n; k++ {
+		if lconst %s r.Values[k] {
+			out.Set(k)
+		}
+	}
+`
+
+const const_idx = `
+	for k := uint32(0); k < n; k++ {
+		if lconst %s r.Values[rx[k]] {
+			out.Set(k)
+		}
+	}
+`
+
+const flat_flat_s = `
+	for k := uint32(0); k < n; k++ {
+		if %s.Compare(l.Value(k), r.Value(k)) %s 0 {
+			out.Set(k)
+		}
+	}
+`
+
+const flat_idx_s = `
+	for k := uint32(0); k < n; k++ {
+		if %s.Compare(l.Value(k), r.Value(uint32(rx[k]))) %s 0 {
+			out.Set(k)
+		}
+	}
+`
+
+const flat_const_s = `
+	for k := uint32(0); k < n; k++ {
+		if %s.Compare(l.Value(k), rconst) %s 0 {
+			out.Set(k)
+		}
+	}
+`
+
+const idx_flat_s = `
+	for k := uint32(0); k < n; k++ {
+		if %s.Compare(l.Value(uint32(lx[k])), r.Value(k)) %s 0 {
+			out.Set(k)
+		}
+	}
+`
+
+const idx_idx_s = `
+	for k := uint32(0); k < n; k++ {
+		if %s.Compare(l.Value(uint32(lx[k])), r.Value(uint32(rx[k]))) %s 0 {
+			out.Set(k)
+		}
+	}
+`
+
+const idx_const_s = `
+	for k := uint32(0); k < n; k++ {
+		if %s.Compare(l.Value(uint32(lx[k])), rconst) %s 0 {
+			out.Set(k)
+		}
+	}
+`
+
+const const_flat_s = `
+	for k := uint32(0); k < n; k++ {
+		if %s.Compare(lconst, r.Value(k)) %s 0 {
+			out.Set(k)
+		}
+	}
+`
+
+const const_idx_s = `
+	for k := uint32(0); k < n; k++ {
+		if %s.Compare(lconst, r.Value(uint32(rx[k]))) %s 0 {
+			out.Set(k)
+		}
+	}
+`
+
+func formToLoop(typ string, lform, rform vector.Form, op string) string {
+	switch typ {
+	case "String", "Bytes":
+		return formToLoops(typ, lform, rform, op)
+	default:
+		return formToLoopv(lform, rform, op)
+	}
+}
+
+func formToLoops(typ string, lform, rform vector.Form, op string) string {
+	pkg := strings.ToLower(typ)
+	if pkg == "string" {
+		pkg = "strings"
+	}
+	switch lform {
+	case vector.FormFlat:
+		switch rform {
+		case vector.FormFlat:
+			return fmt.Sprintf(flat_flat_s, pkg, op)
+		case vector.FormDict:
+			return fmt.Sprintf(flat_idx_s, pkg, op)
+		case vector.FormConst:
+			return fmt.Sprintf(flat_const_s, pkg, op)
+		}
+	case vector.FormDict:
+		switch rform {
+		case vector.FormFlat:
+			return fmt.Sprintf(idx_flat_s, pkg, op)
+		case vector.FormDict:
+			return fmt.Sprintf(idx_idx_s, pkg, op)
+		case vector.FormConst:
+			return fmt.Sprintf(idx_const_s, pkg, op)
+		}
+	case vector.FormConst:
+		switch rform {
+		case vector.FormFlat:
+			return fmt.Sprintf(const_flat_s, pkg, op)
+		case vector.FormDict:
+			return fmt.Sprintf(const_idx_s, pkg, op)
+		}
```

Comment:
For my own edification- there no such thing as having a const on both the lhs and rhs?

Reply:
Right, because in that case it's more efficient to return a vector.Const, and that's what vam/expr.compareConsts does.